### PR TITLE
feat: implement Element.waitForSelector

### PR DESCRIPTION
**Pull request by OrKoN**
Issues: #6427

### CODE REVIEW

File: src/common/JSHandle.ts [hunk 1]
```diff
@@ -351,6 +351,50 @@ export class ElementHandle<
...
+   * {@link https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors | selector}
+   * of an element to wait for
+   * @param options - Optional waiting parameters
+   * @returns Promise which resolves when element specified by selector string
+   * is added to DOM. Resolves to `null` if waiting for hidden: `true` and
+   * selector is not found in DOM.
+   * @remarks
+   * The optional Parameter in Arguments `options` are :
+   *
+   * - `Visible`: A boolean wait for element to be present in DOM and to be
```
Comment (johanbay) [inline src/common/JSHandle.ts]:
```suggestion
   * - `visible`: wait for the selected element to be present in DOM and to be
```

File: src/common/JSHandle.ts [hunk 2]
```diff
@@ -351,6 +351,50 @@ export class ElementHandle<
...
+   *
+   * @param selector - A
+   * {@link https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors | selector}
+   * of an element to wait for
+   * @param options - Optional waiting parameters
+   * @returns Promise which resolves when element specified by selector string
+   * is added to DOM. Resolves to `null` if waiting for hidden: `true` and
+   * selector is not found in DOM.
+   * @remarks
+   * The optional Parameter in Arguments `options` are :
```
Comment (johanbay) [inline src/common/JSHandle.ts]:
```suggestion
   * The optional parameters in `options` are:
```

File: src/common/JSHandle.ts [hunk 3]
```diff
@@ -351,6 +351,50 @@ export class ElementHandle<
...
+   *
+   * - `Visible`: A boolean wait for element to be present in DOM and to be
+   * visible, i.e. to not have `display: none` or `visibility: hidden` CSS
+   * properties. Defaults to `false`.
+   *
+   * - `hidden`: ait for element to not be found in the DOM or to be hidden,
```
Comment (johanbay) [inline src/common/JSHandle.ts]:
```suggestion
   * - `hidden`: wait for the selected element to not be found in the DOM or to be hidden,
```

File: src/common/JSHandle.ts [hunk 4]
```diff
@@ -351,6 +351,50 @@ export class ElementHandle<
...
+   * i.e. have `display: none` or `visibility: hidden` CSS properties. Defaults to
+   * `false`.
+   *
+   * - `timeout`: maximum time to wait for in milliseconds. Defaults to `30000`
```
Comment (johanbay) [inline src/common/JSHandle.ts]:
```suggestion
   * - `timeout`: maximum time to wait in milliseconds. Defaults to `30000`
```

**Review by johanbay — APPROVED** [review]
LGTM % nits

**Comment by OrKoN** [comment]
@mathiasbynens pls take a look!

File: docs/api.md [hunk 5]
```diff
@@ -4872,6 +4873,21 @@ await elementHandle.press('Enter');
...
+- `selector` <[string]> A [selector] of an element to wait for
+- `options` <[Object]> Optional waiting parameters
+  - `visible` <[boolean]> wait for element to be present in DOM and to be visible, i.e. to not have `display: none` or `visibility: hidden` CSS properties. Defaults to `false`.
+  - `hidden` <[boolean]> wait for element to not be found in the DOM or to be hidden, i.e. have `display: none` or `visibility: hidden` CSS properties. Defaults to `false`.
+  - `timeout` <[number]> maximum time to wait for in milliseconds. Defaults to `30000` (30 seconds). Pass `0` to disable timeout. The default value can be changed by using the [page.setDefaultTimeout(timeout)](#pagesetdefaulttimeouttimeout) method.
+- returns: <[Promise]<?[ElementHandle]>> Promise which resolves when element specified by selector string is added to DOM. Resolves to `null` if waiting for `hidden: true` and selector is not found in DOM.
+
+Wait for the `selector` to appear within the element. If at the moment of calling
+the method the `selector` already exists, the method will return
+immediately. If the selector doesn't appear after the `timeout` milliseconds of waiting, the function will throw.
```
Comment (mathiasbynens) [inline docs/api.md]:
```suggestion
Wait for an element matching `selector` to appear within the `elementHandle`’s subtree. If the `selector` already matches an element at the moment of calling the method, the promise returned by the method resolves immediately. If the selector doesn’t appear after `timeout` milliseconds of waiting, the promise rejects.
```

**Review by mathiasbynens — APPROVED** [review]
LGTM % some suggestions to make the docs precise

**Pull request merged**